wl-sim/main/WL_Flash.cpp

```diff
@@ -0,0 +1,292 @@
+        keys[i] = rand() % UINT8_MAX;
+
+    ESP_LOGI(TAG, "%s: generated 8bit keys (%u, %u, %u) ", __func__, keys[0], keys[1], keys[2]);
+
+}
+
+static uint32_t feistel_function(uint32_t msb, uint32_t key)
+{
+    ESP_LOGV(TAG, "%s: msb=0x%x key=0x%x", __func__, msb, key);
+    ESP_LOGV(TAG, "%s: msb xor key = 0x%x", __func__, (msb ^ key));
+    ESP_LOGV(TAG, "%s: return (msb xor key)^2 = 0x%x", __func__, (msb ^ key)*(msb ^ key));
+    return (msb ^ key) * (msb ^ key);
+}
+
+size_t feistel_network(size_t logical_addr)
+{
+    feistel_calls++;
+    size_t addr = logical_addr;
+
+round:
+    size_t sector_addr = addr / SECTOR_SIZE;
+
+    //               |       B       |
+    //               |<-MSB->|<-LSB->|
+    // logical_addr: |  msb  |  lsb  |
+
+    size_t B_mask = ~( (~(size_t)0) << B );
+
+    size_t LSB_mask = ~( (~(size_t)0) << LSB );
+    size_t MSB_mask = ( (~(size_t)0) << LSB ) & B_mask;
+
+    size_t msb, lsb, _msb, _lsb, randomized_addr;
+
+    for (auto i = 0; i < 3; i++) {
+        msb = sector_addr >> LSB;
+        lsb = sector_addr & LSB_mask;
+
+        _msb = msb;
+        // mask output of F to also be |LSB| for XORing with lsb
+        _lsb = (lsb ^ (feistel_function(msb, keys[i]) & LSB_mask));
+
+        // swap lsb and msb
+        sector_addr = (_lsb << MSB) | _msb;
+        ESP_LOGV(TAG, "%s: msb=0x%x lsb=0x%x sector_addr=0x%x", __func__, _msb, _lsb, sector_addr);
+    }
+    randomized_addr = sector_addr;
+    ESP_LOGD(TAG, "%s: randomized_addr=0x%x", __func__, randomized_addr);
+
+    if (randomized_addr >= SECTOR_COUNT) {
+        addr = randomized_addr * SECTOR_SIZE;
+        feistel_cycle_walks++;
+#if 0
+        ESP_LOGE(TAG, "%s: logical=0x%x => randomized 0x%x outside of domain 0x%x, cycle walk", __func__, logical_addr, randomized_addr, SECTOR_COUNT);
+
+        if (feistel_cycle_walks > 20) {
+            ESP_LOGE(TAG, "%s: feistel cycle walks critical, exiting", __func__);
+            exit(1);
+        }
+#endif
+        goto round;
+    }
+
+    return randomized_addr * SECTOR_SIZE;
+}
+
+size_t calcAddr(size_t addr)
+{
+    size_t intermediate_addr = addr;
+    if (feistel)
+        intermediate_addr = feistel_network(addr);
+
+    size_t result = (FLASH_SIZE - move_count * PAGE_SIZE + intermediate_addr) % FLASH_SIZE;
+    size_t dummy_addr = pos * PAGE_SIZE;
+
+    if (result < dummy_addr) {
+    } else {
+        result += PAGE_SIZE;
+    }
+
+    ESP_LOGV(TAG, "%s - addr= 0x%08x -> result= 0x%08x, dummy_addr= 0x%08x", __func__, (uint32_t) addr, (uint32_t) result, (uint32_t)dummy_addr); 
+    return result;
+}
+
+esp_err_t updateWL()
+{
+    esp_err_t result = ESP_OK;
+
+    access_count++;
+    if (access_count < MAX_COUNT) {
+        ESP_LOGV(TAG, "%s EARLY RETURN - access_count= 0x%08x, pos= 0x%08x, move_count= 0x%08x", __func__, (uint32_t) access_count, (uint32_t) pos, (uint32_t) move_count); 
+        return result;
+    }
+
+    access_count = 0;
+
+    // copy data to dummy block
```

Comment:
I assume this is to be updated - generally, comments should be brief and descriptive, providing desired information at the first glance

Reply:
removed misleading comments